src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/CoverageModelEMComputeBlock.java

```diff
@@ -0,0 +1,1332 @@
+ * TODO github/gatk-protected issue #853 -- use instrumentation to measure memory consumption
+
+ * @implNote Methods that manipulate INDArrays must make sure to leave queried values from
+ * {@link CoverageModelEMComputeBlock#icg} unchanged. For example, to calculate the A.B.C + D,
+ * the correct template is first fetch the data from ICG as follows:
+ *
+ * // query cached values
+ * final INDArray A = getINDArrayFromCache(CoverageModelICGCacheNode.A);
+ * final INDArray B = getINDArrayFromCache(CoverageModelICGCacheNode.B);
+ * final INDArray C = getINDArrayFromCache(CoverageModelICGCacheNode.C);
+ * final INDArray D = getINDArrayFromCache(CoverageModelICGCacheNode.D);
+ *
+ * and then call:
+ *
+ * // calculations
+ * final INDArray result_1 = A.mul(B).muli(C).addi(D); // OK
+ *                           ---1--- (a copy is instantiated; call it tmp)
+ *                                ----2---  (tmp is modified in-place)
+ *                                    -----3---- (tmp is modified in-place)
+ *
+ * (note the difference between mul(...) and muli(...)
+ *
+ * Here, A and multiplied by B to create a new INDArray; subsequent operations can be performed
+ * in-place to avoid unnecessary memory allocation/deallocation and garbage generation.
+ *
+ * It is, however, illegal to perform the following operation:
+ *
+ * final INDArray result_3 = A.muli(B).muli(C).addi(D); // NOT OK
+ *                           ---1--- (A is modified in-place!)
+ *
+ * Here, A is modified in-place which corrupts the data in {@link CoverageModelEMComputeBlock#icg}
+ *
+ * The code reviewer must CAREFULLY check all of the matrix computations to ensure that the such
+ * case does not occur, and suggest optimizations if noted.
+ *
+ * Also, since we do not have an immutable INDArray (or a wrapper for that),
+ * it is the user's responsibility to ensure that the queried INDArray's are not mutated.
+ * The safest approach is to call dup() on cached queries, however, it leads to
+ * poor memory performance.
+ *
+ * @author Mehrtash Babadi &lt;mehrtash@broadinstitute.org&gt;
+ */
+public final class CoverageModelEMComputeBlock {
+
+    /**
+     * An instance of {@link ImmutableComputableGraph} for automatic evaluation, caching, and bookkeeping
+     * of most evaluations
+     */
+    private final ImmutableComputableGraph icg;
+
+    /**
+     * The index range of targets included in this compute block
+     */
+    private final LinearSpaceBlock targetBlock;
+
+    private final int numSamples, numLatents, numTargets;
+
+    /**
+     * The latest signal from calling an M-step subroutine
+     */
+    private final SubroutineSignal latestMStepSignal;
+
+    /**
+     * Immutable computable graph cache nodes
+     */
+    public enum CoverageModelICGCacheNode {
+        n_st("Read count"),
+        M_st("Learning mask (only 0 and 1 values)"),
+        MER_st("Mapping error rate"),
```

Comment:
It is a property of both in practice, right? If a sample has a lot of low quality reads that overlap with a target, the mapping error rate will be high (because the same low quality reads can be poorly aligned to many places).